### PR TITLE
Receive thrift function calls from iOS and Android

### PR DESCRIPTION
**Pull request by webb04**
## Why are you doing this?
Some tasks will require the native layer to pass a message to the webview and will expect an action and/or response.

An example of this is inserting epics into the html.

### CODE REVIEW

**Comment by alexanderedge** [comment]
Max raised a question related to this which I posted into the mobile-apps-thrift repo https://github.com/guardian/mobile-apps-thrift/issues/12.

File: src/native/thrift/nativeConnection.ts [hunk 1]
```diff
@@ -123,14 +128,45 @@ export class NativeConnection<Context = void> extends ThriftConnection {
     }
 }
 
-function receiveiOSMessage(buffer: Buffer): void {
-    const transport = new BufferedTransport(buffer);
-    const protocol = new CompactProtocol(transport);
-    const processor = new Webview.Processor(new WebviewHandler);
-    processor.process(protocol, protocol, null);
+class ThriftServer<T> {
```
Comment (davidfurey) [inline src/native/thrift/nativeConnection.ts]: Should this be a different file?

File: src/native/thrift/nativeConnection.ts [hunk 2]
```diff
@@ -123,14 +128,45 @@ export class NativeConnection<Context = void> extends ThriftConnection {
...
+    }
+
+    constructor(processor: ThriftProcessor<null, T>) {
+        this.processor = processor
+        window.receiveNativeMessage = this.receive;
+    }
 }
 
-window.receiveiOSMessage = receiveiOSMessage;
+new ThriftServer(new Webview.Processor(new WebviewHandler));
```
Comment (davidfurey) [inline src/native/thrift/nativeConnection.ts]: This should probably live somewhere else

File: src/native/thrift/nativeConnection.ts [hunk 3]
```diff
@@ -21,15 +22,19 @@ import { WebviewHandler } from 'native/webviewApi';
 declare global {
     interface Window {
         nativeConnections: { [id: string]: NativeConnection };
-        AndroidWebViewMessage?: (nativeMessage: NativeMessage) => {};
+        AndroidWebViewRequest?: (nativeMessage: NativeMessage) => {};
```
Comment (davidfurey) [inline src/native/thrift/nativeConnection.ts]: What do you think about changing these names?
`AndroidWebViewRequest?`
`AndroidNativeResponse?`

`iOSWebViewRequest`
`iOSNativeResponse`

`receiveNativeRequest`

File: src/native/thrift/nativeConnection.ts [hunk 4]
```diff
@@ -94,7 +89,17 @@ export class NativeConnection<Context = void> extends ThriftConnection {
         const message = this.outBuffer.shift();
         if (message) {
             console.log("Sending next message")
```
Comment (davidfurey) [inline src/native/thrift/nativeConnection.ts]: Should we remove this or change it to `console.debug`?

File: src/native/thrift/nativeConnection.ts [hunk 5]
```diff
@@ -9,22 +9,32 @@ import {
...
-                iOSWebViewMessage: {
+                iOSWebViewRequest: {
+                    postMessage: (nativeMessage: NativeMessage) => {};
+                };
+                iOSWebViewResponse: {
                     postMessage: (nativeMessage: NativeMessage) => {};
                 };
             };
         };
+        receiveNativeMessage: (nativeMessage: NativeMessage) => void;
```
Comment (webb04) [inline src/native/thrift/nativeConnection.ts]: Should `NativeMessage` be renamed to `Message` as it's used both ways now. Maybe the `connectionId` should be optional or we have two separate types?

Comment (davidfurey) [reply src/native/thrift/nativeConnection.ts]: I think the connection id should always be required. Probably makes sense to rename to `Message` or maybe `ThriftMessage`?

File: src/native/thrift/nativeConnection.ts (hunk 4 above)
Comment (webb04) [reply src/native/thrift/nativeConnection.ts]: We shouldn't send them to Prod but maybe we look at adding https://stackoverflow.com/questions/41040266/remove-console-logs-with-webpack-uglify

**Comment by webb04** [comment]
Closing this PR as it makes sense to add the changes directly onto the new viewController.

This PR implemented a two way thrift communication:

native calling webview
webview calling native

It can be used as a reference with this [iOS](https://github.com/guardian/ios-live/pull/4940).

**Pull request closed without merging**